### PR TITLE
PF-1174: Add reference type bucket file

### PR DESCRIPTION
**Pull request by yuhuyoyo**
This is a draft. Starting the github actions.

### CODE REVIEW

File: service/src/main/java/bio/terra/workspace/app/controller/ReferencedGcpResourceController.java [hunk 1]
```diff
@@ -67,8 +71,70 @@ private AuthenticatedUserRequest getAuthenticatedInfo() {
     return authenticatedUserRequestFactory.from(request);
   }
 
-  // -- GCS Bucket -- //
+  // -- GSC Bucket file -- //
+  @Override
+  public ResponseEntity<ApiGcpGcsBucketFileResource> createBucketFileReference(
+      UUID id, @Valid ApiCreateGcpGcsBucketFileReferenceRequestBody body) {
```
Comment (jaycarlton) [inline service/src/main/java/bio/terra/workspace/app/controller/ReferencedGcpResourceController.java:77]: we usually rename this `id` to `workspaceId` for clarity.

File: service/src/main/java/bio/terra/workspace/service/resource/referenced/ReferencedGcsBucketFileResource.java [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+   * @param description description - may be null
+   * @param cloningInstructions cloning instructions
+   * @param bucketName bucket name
+   * @param fileName name for the file in the bucket
+   */
+  @JsonCreator
+  public ReferencedGcsBucketFileResource(
+      @JsonProperty("workspaceId") UUID workspaceId,
+      @JsonProperty("resourceId") UUID resourceId,
+      @JsonProperty("name") String name,
```
Comment (jaycarlton) [inline service/src/main/java/bio/terra/workspace/service/resource/referenced/ReferencedGcsBucketFileResource.java:41]: add `@Nullable` to the nullable ones

File: service/src/main/resources/api/service_openapi.yaml [hunk 3]
```diff
@@ -2965,6 +3111,31 @@ components:
...
+        Bucket file properties included in post-creation, get, and update. Others must be
+        retrieved from GCS using the name.
+      type: object
+      required: [ bucketName, fileName ]
+      properties:
+        bucketName:
+          description: Name of created bucket (not the resource name).
+          type: string
+        fileName:
+          description: Name of the file in the gcs bucket.
```
Comment (jaycarlton) [inline service/src/main/resources/api/service_openapi.yaml:3125]: Maybe indicate this is the full path to the file in the bucket.

File: service/src/test/java/bio/terra/workspace/service/resource/referenced/ReferencedResourceServiceTest.java [hunk 4]
```diff
@@ -399,6 +399,53 @@ void setup() throws Exception {
...
+            referenceResourceService.checkAccess(
+                workspaceId, referenceResource.getResourceId(), USER_REQUEST));
+
+        ReferencedResource byid =
+            referenceResourceService.getReferenceResource(
+                workspaceId, resource.getResourceId(), USER_REQUEST);
+        ReferencedResource byname =
+            referenceResourceService.getReferenceResourceByName(
+                workspaceId, resource.getName(), USER_REQUEST);
+        assertEquals(byid.castToGcsBucketFileResource(), byname.castToGcsBucketFileResource());
```
Comment (jaycarlton) [inline service/src/test/java/bio/terra/workspace/service/resource/referenced/ReferencedResourceServiceTest.java:440]: maybe assert that these aren't both null

**Review by jaycarlton — APPROVED** [review]
Looks good!

**Review by ddietterich — APPROVED** [review]
LGTM

**Pull request merged**